src/algorand_transactions/transactions_group.rs

```diff
@@ -0,0 +1,246 @@
+use serde::{Deserialize, Serialize};
+use sha2::Digest;
+
+use crate::{
+    algorand_errors::AlgorandError,
+    algorand_types::{Byte, Bytes, Result},
+    AlgorandHash,
+    AlgorandKeys,
+    AlgorandSignedTransaction,
+    AlgorandTransaction,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct TxGroup {
+    #[serde(rename = "txlist", default)]
+    tx_group_hashes: Vec<AlgorandHash>,
+
+    #[serde(skip_serializing)]
+    txs: Vec<AlgorandTransaction>,
+
+    #[serde(skip_serializing)]
+    group_id: AlgorandHash,
+}
+
+impl TxGroup {
+    const MAX_TX_GROUP_SIZE: usize = 16;
+
+    pub fn new(transactions: &[AlgorandTransaction]) -> Result<TxGroup> {
+        transactions
+            .iter()
+            .map(|tx| tx.to_raw_tx_id())
+            .collect::<Result<Vec<AlgorandHash>>>()
+            .and_then(|raw_tx_ids| {
+                let tx_group = Self {
+                    group_id: AlgorandHash::default(),
+                    txs: vec![],
+                    tx_group_hashes: raw_tx_ids.clone(),
+                };
+
+                let group_id = tx_group.compute_group_id()?;
+
+                let txs_with_group_ids_assigned = transactions
+                    .iter()
+                    .map(|tx| tx.assign_group_id(group_id))
+                    .collect::<Vec<AlgorandTransaction>>();
+
+                Ok(Self {
+                    group_id,
+                    tx_group_hashes: raw_tx_ids,
+                    txs: txs_with_group_ids_assigned,
+                })
+            })
+    }
+
+    pub fn compute_group_id(&self) -> Result<AlgorandHash> {
+        if self.tx_group_hashes.is_empty() {
+            return Err(AlgorandError::Custom("Empty transactions list".to_string()));
```

Comment:
Not that it requires changing, but as I mentioned elsewhere, we can do exactly the same thing but more succinctly like so:

```suggestion
            return Err("Empty transactions list".into());
```

The compiler will figure out it's for the `Custom` variant.